### PR TITLE
[Merged by Bors] - Improve safety for `CommandQueue` internals

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

- Safety comments for the `CommandQueue` type are quite sparse and very imprecise. Sometimes, they are right for the wrong reasons or use circular reasoning.

## Solution

- Document previously-implicit safety invariants.
- Rewrite safety comments to actually reflect the specific invariants of each operation.
- Use `OwningPtr` instead of raw pointers, to encode an invariant in the type system instead of via comments.
- Use typed pointer methods when possible to increase reliability.

---

## Changelog

+ Added the function `OwningPtr::read_unaligned`.

### CODE REVIEW

**Review by james7132 — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/system/commands/command_queue.rs [hunk 1]
```diff
@@ -34,45 +43,41 @@ impl CommandQueue {
...
-            // later, a double `drop` does not occur.
-            unsafe {
-                std::ptr::copy_nonoverlapping(
-                    &*command as *const C as *const MaybeUninit<u8>,
-                    self.bytes.as_mut_ptr().add(old_len),
-                    size,
-                );
-                self.bytes.set_len(old_len + size);
-            }
+            // SAFETY: Due to the call to `.reserve()` above, the buffer must be
```
[the remark this answers is not in the document]
Comment (JoJoJet) [reply crates/bevy_ecs/src/system/commands/command_queue.rs:65]: Good point.

File: crates/bevy_ptr/src/lib.rs [hunk 2]
```diff
@@ -214,16 +214,24 @@ impl<'a> OwningPtr<'a> {
     /// Consumes the [`OwningPtr`] to obtain ownership of the underlying data of type `T`.
     ///
     /// # Safety
-    /// Must point to a valid `T`.
+    /// Must point to an aligned and fully-initialized value of type `T`.
```
[the remark this answers is not in the document]
Comment (JoJoJet) [reply crates/bevy_ptr/src/lib.rs:217]: Aren't you adding this in #7117?

File: crates/bevy_ecs/src/system/commands/command_queue.rs [hunk 3]
```diff
@@ -1,23 +1,32 @@
...
 
 use super::Command;
 use crate::world::World;
 
 struct CommandMeta {
+    /// Offset from the start of `CommandQueue.bytes` at which the corresponding command is stored.
     offset: usize,
-    func: unsafe fn(value: *mut MaybeUninit<u8>, world: &mut World),
+    /// SAFETY: The `value` must point to a value of type `T: Command`,
+    /// where `T` is some specific type that was used to produce this metadata.
```
[the remark this answers is not in the document]
Comment (JoJoJet) [reply crates/bevy_ecs/src/system/commands/command_queue.rs:12]: The function has no requirements about alignment, so there's no need to mention it.

**Review by alice-i-cecile — APPROVED** [review]
This is a dramatic improvement in clarity.

File: crates/bevy_ecs/src/system/commands/command_queue.rs [hunk 4]
```diff
@@ -34,45 +44,45 @@ impl CommandQueue {
...
-            let command = command.cast::<T>().read_unaligned();
-            command.write(world);
-        }
-
-        let size = std::mem::size_of::<C>();
         let old_len = self.bytes.len();
 
+        // SAFETY: After adding the metadata, we correctly write the corresponding `command`
+        // of type `C` into `self.bytes`. Zero-sized commands do not get written into the buffer,
+        // so we'll just use a dangling pointer, which is always valid for zero-sized types.
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/system/commands/command_queue.rs:51]:
```suggestion
        // so we'll just use a dangling pointer, which is valid for zero-sized types.
```

File: crates/bevy_ptr/src/lib.rs [hunk 5]
```diff
@@ -328,6 +328,15 @@ impl<'a, A: IsAligned> OwningPtr<'a, A> {
         unsafe { PtrMut::new(self.0) }
     }
 }
+impl<'a> OwningPtr<'a, Unaligned> {
```
Comment (james7132) [inline crates/bevy_ptr/src/lib.rs:331]:
```suggestion

impl<'a> OwningPtr<'a, Unaligned> {
```

**Review by james7132 — APPROVED** [review]

**Pull request closed without merging**